File: HealthApp/Tabs/History/ViewModel/HistoryViewModel.m
//
//  HistoryViewModel.m
//  HealthApp
//
//  Created by Christopher Ray on 3/27/21.
//

#import "HistoryViewModel.h"
#import "PersistenceService.h"
#import "AppUserData.h"
#include "CalendarDateHelpers.h"
#import "ChartHelpers.h"
#import "WeeklyData+CoreDataClass.h"
#include "array.h"

#define freeChartDataEntry(x) [(x) release]
#define SmallDataSetCutoff 7

gen_array(chartData, ChartDataEntry*, DSDefault_shallowCopy, freeChartDataEntry)

typedef struct {
    double weekStart;
    double weekEnd;
    int totalWorkouts;
    int durationByType[4];
    int cumulativeDuration[4];
    int weightArray[4];
} WeekDataModel;

gen_array(weekData, WeekDataModel, DSDefault_shallowCopy, DSDefault_shallowDelete)

HistoryViewModel *historyViewModel_init(void) {
    HistoryViewModel *model = malloc(sizeof(HistoryViewModel));
    if (!model) return NULL;
    model->areaChartViewModel = calloc(1, sizeof(HistoryAreaChartViewModel));
    model->liftChartViewModel = calloc(1, sizeof(HistoryLiftChartViewModel));
    model->gradientChartViewModel = calloc(1, sizeof(HistoryGradientChartViewModel));
    if (!(model->areaChartViewModel && model->gradientChartViewModel && model->liftChartViewModel)) {
        if (model->areaChartViewModel) free(model->areaChartViewModel);
        if (model->gradientChartViewModel) free(model->gradientChartViewModel);
        if (model->liftChartViewModel) free(model->liftChartViewModel);
        free(model);
        return NULL;
    }

    {
        HistoryAreaChartViewModel *vm = model->areaChartViewModel;
        for (int i = 0; i < 5; ++i) {
            vm->entries[i] = array_new(chartData);
        }
        vm->legendLabelFormats[0] = CFStringCreateWithCString(NULL, "Strength (Avg: %@)", kCFStringEncodingUTF8);
        vm->legendLabelFormats[1] = CFStringCreateWithCString(NULL, "HIC (Avg: %@)", kCFStringEncodingUTF8);
        vm->legendLabelFormats[2] = CFStringCreateWithCString(NULL, "SE (Avg: %@)", kCFStringEncodingUTF8);
        vm->legendLabelFormats[3] = CFStringCreateWithCString(NULL, "Endurance (Avg: %@)", kCFStringEncodingUTF8);
    }
    {
        HistoryLiftChartViewModel *vm = model->liftChartViewModel;
        for (int i = 0; i < 4; ++i) {
            vm->entries[i] = array_new(chartData);
        }
        vm->legendLabelFormats[0] = CFStringCreateWithCString(NULL, "Squat (Avg: %.1f)", kCFStringEncodingUTF8);
        vm->legendLabelFormats[1] = CFStringCreateWithCString(NULL, "Pull-up (Avg: %.1f)", kCFStringEncodingUTF8);
        vm->legendLabelFormats[2] = CFStringCreateWithCString(NULL, "Bench (Avg: %.1f)", kCFStringEncodingUTF8);
        vm->legendLabelFormats[3] = CFStringCreateWithCString(NULL, "Deadlift (Avg: %.1f)", kCFStringEncodingUTF8);
    }
    {
        HistoryGradientChartViewModel *vm = model->gradientChartViewModel;
        vm->legendLabelFormat = CFStringCreateWithCString(NULL, "Avg Workouts (%.2f)", kCFStringEncodingUTF8);
        vm->entries = array_new(chartData);
    }
    model->data = NULL;
    return model;
}

void historyViewModel_free(HistoryViewModel *model) {
    {
        HistoryAreaChartViewModel *vm = model->areaChartViewModel;
        for (int i = 0; i < 4; ++i) { CFRelease(vm->legendLabelFormats[i]); }
        for (int i = 0; i < 5; ++i) {  array_free(chartData, vm->entries[i]); }
        free(vm);
    }
    {
        HistoryLiftChartViewModel *vm = model->liftChartViewModel;
        for (int i = 0; i < 4; ++i) { CFRelease(vm->legendLabelFormats[i]); }
        for (int i = 0; i < 4; ++i) {  array_free(chartData, vm->entries[i]); }
        free(vm);
    }
    {
        HistoryGradientChartViewModel *vm = model->gradientChartViewModel;
        CFRelease(vm->legendLabelFormat);
        array_free(chartData, vm->entries);
        free(vm);
    }
    free(model);
}

void historyViewModel_fetchData(HistoryViewModel *model) {
    CFCalendarRef calendar = CFCalendarCopyCurrent();

    if (!model->data) {
        model->data = array_new(weekData);
        array_reserve(weekData, model->data, 128);
    } else {
        array_clear(weekData, model->data);
    }

    size_t count = 0;
    NSFetchRequest *fetchRequest = WeeklyData.fetchRequest;
    fetchRequest.predicate = [NSPredicate predicateWithFormat:@"weekStart > %f AND weekStart < %f", date_twoYears(calendar), appUserDataShared->weekStart - 2];
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"weekStart" ascending:true];
    fetchRequest.sortDescriptors = @[descriptor];
    [descriptor release];
    CFRelease(calendar);

    NSArray<WeeklyData *> *data = [persistenceService_sharedContainer.viewContext executeFetchRequest:fetchRequest error:nil];
    if (!(data && (count = (data.count)) != 0)) return;

    for (size_t i = 0; i < count; ++i) {
        WeeklyData *d = data[i];
        WeekDataModel m = {
            .weekStart = d.weekStart, .weekEnd = d.weekEnd, .totalWorkouts = d.totalWorkouts, .weightArray = {d.bestSquat, d.bestPullup, d.bestBench, d.bestDeadlift}, .durationByType = {d.timeStrength, d.timeHIC, d.timeSE, d.timeEndurance}
        };

        m.cumulativeDuration[0] = m.durationByType[0];
        m.cumulativeDuration[1] = m.cumulativeDuration[0] + m.durationByType[1];
        m.cumulativeDuration[2] = m.cumulativeDuration[1] + m.durationByType[2];
        m.cumulativeDuration[3] = m.cumulativeDuration[2] + m.durationByType[3];

        array_push_back(weekData, model->data, m);
    }
}

void historyViewModel_formatDataForTimeRange(HistoryViewModel *model, int index) {
    {
        HistoryAreaChartViewModel *vm = model->areaChartViewModel;
        vm->maxActivityTime = 0;
        memset(vm->totalByType, 0, 4 * sizeof(int));
        for (int i = 0; i < 5; ++i) {
            array_clear(chartData, vm->entries[i]);
        }
    }
    {
        HistoryLiftChartViewModel *vm = model->liftChartViewModel;
        vm->maxWeight = 0;
        memset(vm->totalByExercise, 0, 4 * sizeof(int));
        for (int i = 0; i < 4; ++i) {
            array_clear(chartData, vm->entries[i]);
        }
    }
    {
        HistoryGradientChartViewModel *vm = model->gradientChartViewModel;
        vm->avgWorkouts = vm->maxWorkouts = vm->totalWorkouts = 0;
        array_clear(chartData, vm->entries);
    }

    int size = 0;
    if (!(size = (array_size(model->data)))) return;

    int startIndex = 0;
    if (index == 0) {
        startIndex = size - 26;
    } else if (index == 1) {
        startIndex = size - 52;
    }

    if (startIndex < 0) startIndex = 0;

    WeekDataModel *arr = model->data->arr;
    double referenceTime = arr[startIndex].weekStart;
    NSString *format = (size - startIndex) < 7 ? @"MMM dd" : @"M/d/yy";
    [sharedHistoryXAxisFormatter update:referenceTime dateFormat:format];

    for (int i = startIndex; i < size; ++i) {
        WeekDataModel *e = &arr[i];
        double xValue = (e->weekStart - referenceTime) / DaySeconds;

        {
            HistoryGradientChartViewModel *viewModel = model->gradientChartViewModel;
            int workouts = e->totalWorkouts;
            viewModel->totalWorkouts += workouts;
            if (workouts > viewModel->maxWorkouts) viewModel->maxWorkouts = workouts;

            array_push_back(chartData, viewModel->entries, [[ChartDataEntry alloc] initWithX:xValue y:workouts]);
        }
        {
            HistoryAreaChartViewModel *viewModel = model->areaChartViewModel;
            for (int i = 0; i < 4; ++i) {
                viewModel->totalByType[i] += e->durationByType[i];
            }

            if (e->cumulativeDuration[3] > viewModel->maxActivityTime) viewModel->maxActivityTime = e->cumulativeDuration[3];

            array_push_back(chartData, viewModel->entries[0], [[ChartDataEntry alloc] initWithX:xValue y:0]);
            for (int i = 1; i < 5; ++i) {
                array_push_back(chartData, viewModel->entries[i], [[ChartDataEntry alloc] initWithX:xValue y:e->cumulativeDuration[i - 1]]);
            }
        }
        {
            HistoryLiftChartViewModel *viewModel = model->liftChartViewModel;
            for (int i = 0; i < 4; ++i) {
                int w = e->weightArray[i];
                viewModel->totalByExercise[i] += w;
                if (w > viewModel->maxWeight) viewModel->maxWeight = w;
                array_push_back(chartData, viewModel->entries[i], [[ChartDataEntry alloc] initWithX:xValue y:w]);
            }
        }
    }

    // finish sub-view model calculations
    {
        HistoryGradientChartViewModel *viewModel = model->gradientChartViewModel;
        viewModel->avgWorkouts = (double) viewModel->totalWorkouts / (size - startIndex);
    }
}

bool historyViewModel_shouldShowCharts(HistoryViewModel *model) {
    return model->data->size > 0;
}

void historyViewModel_applyUpdatesForTotalWorkouts(HistoryGradientChartViewModel *model, LineChartView *view, LineChartData *data,
                                                   LineChartDataSet *dataSet, ChartLimitLine *limitLine, NSArray<ChartLegendEntry*> *legendEntries) {
    int entryCount = array_size(model->entries);
    bool isSmallDataSet = entryCount < SmallDataSetCutoff;
    dataSet.drawCirclesEnabled = isSmallDataSet;
    [dataSet replaceEntries:[NSArray arrayWithObjects:model->entries->arr count:entryCount]];
    [data setDrawValues:isSmallDataSet];

    view.leftAxis.axisMaximum = max(1.1 * model->maxWorkouts, 7);
    view.xAxis.forceLabelsEnabled = isSmallDataSet;
    view.xAxis.labelCount = isSmallDataSet ? entryCount : (SmallDataSetCutoff - 1);
    limitLine.limit = model->avgWorkouts;

    CFStringRef label = CFStringCreateWithFormat(NULL, NULL, model->legendLabelFormat, model->avgWorkouts);
    legendEntries[0].label = (__bridge NSString*) label;
    CFRelease(label);
    view.legend.enabled = true;

    view.data = data;
    [view.data notifyDataChanged];
    [view notifyDataSetChanged];
    [view animateWithXAxisDuration:isSmallDataSet ? 1.5 : 2.5];
}

void historyViewModel_applyUpdatesForDurations(HistoryAreaChartViewModel *model, LineChartView *view, LineChartData *data,
                                               LineChartDataSet **dataSets, NSArray<ChartLegendEntry*> *legendEntries) {
    int entryCount = array_size(model->entries[0]);
    bool isSmallDataSet = entryCount < SmallDataSetCutoff;

    [dataSets[0] replaceEntries:[NSArray arrayWithObjects:model->entries[0]->arr count:entryCount]];
    for (int i = 1; i < 5; ++i) {
        dataSets[i].drawCirclesEnabled = isSmallDataSet;
        [dataSets[i] replaceEntries:[NSArray arrayWithObjects:model->entries[i]->arr count:entryCount]];

        int average = model->totalByType[i - 1] / entryCount;
        CFStringRef suffix;
        if (average > 59) {
            suffix = CFStringCreateWithFormat(NULL, NULL, CFSTR("%d h %d m"), average / 60, average % 60);
        } else {
            suffix = CFStringCreateWithFormat(NULL, NULL, CFSTR("%d m"), average);
        }

        CFStringRef label = CFStringCreateWithFormat(NULL, NULL, model->legendLabelFormats[i - 1], suffix);
        legendEntries[i - 1].label = (__bridge NSString*) label;
        CFRelease(label);
        CFRelease(suffix);
    }
    [data setDrawValues:isSmallDataSet];

    view.leftAxis.axisMaximum = 1.1 * model->maxActivityTime;
    view.xAxis.forceLabelsEnabled = isSmallDataSet;
    view.xAxis.labelCount = isSmallDataSet ? entryCount : (SmallDataSetCutoff - 1);
    view.legend.enabled = true;

    view.data = data;
    [view.data notifyDataChanged];
    [view notifyDataSetChanged];
    [view animateWithXAxisDuration:isSmallDataSet ? 1.5 : 2.5];
}

void historyViewModel_applyUpdatesForLifts(HistoryLiftChartViewModel *model, LineChartView *view, LineChartData *data,
                                           LineChartDataSet **dataSets, NSArray<ChartLegendEntry*> *legendEntries) {
    int entryCount = array_size(model->entries[0]);
    bool isSmallDataSet = entryCount < SmallDataSetCutoff;

    for (int i = 0; i < 4; ++i) {
        dataSets[i].drawCirclesEnabled = isSmallDataSet;
        [dataSets[i] replaceEntries:[NSArray arrayWithObjects:model->entries[i]->arr count:entryCount]];
        double average = (double) model->totalByExercise[i] / (double) entryCount;
        CFStringRef label = CFStringCreateWithFormat(NULL, NULL, model->legendLabelFormats[i], average);
        legendEntries[i].label = (__bridge NSString*) label;
        CFRelease(label);
    }
    [data setDrawValues:isSmallDataSet];

    view.leftAxis.axisMaximum = 1.1 * model->maxWeight;
    view.xAxis.forceLabelsEnabled = isSmallDataSet;
    view.xAxis.labelCount = isSmallDataSet ? entryCount : (SmallDataSetCutoff - 1);
    view.legend.enabled = true;

    view.data = data;
    [view.data notifyDataChanged];
    [view notifyDataSetChanged];
    [view animateWithXAxisDuration:isSmallDataSet ? 1.5 : 2.5];
}

File: HealthApp/Tabs/History/ViewModel/HistoryViewModel.c
//
//  HistoryViewModel.c
//  HealthApp
//
//  Created by Christopher Ray on 3/27/21.
//

#include "HistoryViewModel.h"
#include "AppUserData.h"
#include "CalendarDateHelpers.h"
#include "Exercise.h"
#include "PersistenceService.h"
#include "SwiftBridging.h"

typedef enum {
    HistoryTimeRange6Months,
    HistoryTimeRange1Year
} HistoryTimeRange;

static inline void createNewEntry(Array_object *arr, int x, int y) {
    array_push_back(object, arr, createChartEntry(x, y));
}

gen_array_source(weekData, HistoryWeekDataModel, DSDefault_shallowCopy, DSDefault_shallowDelete)

void historyViewModel_init(HistoryViewModel *this) {
    id chartColors[4];
    for (int i = 0; i < 4; ++i) {
        CFStringRef name = CFStringCreateWithFormat(NULL, NULL, CFSTR("chartColor%d"), i);
        chartColors[i] = staticMethodWithString(objc_getClass("UIColor"),
                                                sel_getUid("colorNamed:"), name);
        CFRelease(name);
    }
    id *areaDataSets = this->workoutTypeModel.dataSets;
    setupLegendEntries(this->totalWorkoutsModel.legendEntries,
                       (id []){createColor("systemTealColor")}, 1);
    setupLegendEntries(this->workoutTypeModel.legendEntries, chartColors, 4);
    setupLegendEntries(this->liftModel.legendEntries, chartColors, 4);

    this->totalWorkoutsModel.entries = array_new(object);
    this->totalWorkoutsModel.dataSet = createDataSet(createColor("systemRedColor"));
    this->workoutTypeModel.dataSets[0] = createEmptyDataSet();
    fillStringArray(this->workoutTypeModel.names, CFSTR("workoutTypes%d"), 4);
    fillStringArray(this->liftModel.names, CFSTR("liftTypes%d"), 4);
    fillStringArray(this->formatter.months, CFSTR("months%02d"), 12);
    this->totalWorkoutsModel.legendFormat = localize(CFSTR("totalWorkoutsLegend"));
    this->liftModel.legendFormat = localize(CFSTR("liftLegend"));
    this->workoutTypeModel.legendFormat = localize(CFSTR("workoutTypeLegend"));

    for (int i = 0; i < 4; ++i) {
        this->workoutTypeModel.entries[i] = array_new(object);
        this->liftModel.entries[i] = array_new(object);

        this->workoutTypeModel.dataSets[i + 1] = createDataSet(chartColors[i]);
        this->liftModel.dataSets[i] = createDataSet(chartColors[i]);
    }

    this->workoutTypeModel.entries[4] = array_new(object);

    this->totalWorkoutsModel.chartData = createChartData((id []){
        this->totalWorkoutsModel.dataSet
    }, 1);
    this->workoutTypeModel.chartData = createChartData((id []){
        areaDataSets[4], areaDataSets[3], areaDataSets[2], areaDataSets[1]
    }, 4);
    this->liftModel.chartData = createChartData(this->liftModel.dataSets, 4);

    this->workoutTypeModel.durationStr = CFStringCreateCopy(NULL, CFSTR(""));
    this->formatter.currString = CFStringCreateCopy(NULL, CFSTR(""));

    this->data = array_new(weekData);
    array_reserve(weekData, this->data, 128);
}

void historyViewModel_fetchData(HistoryViewModel *this) {
    runInBackground((^{
        array_clear(weekData, this->data);
        struct tm localInfo;
        int count = 0;

        id request = fetchRequest(createPredicate(CFSTR("weekStart < %lld"), userData->weekStart));
        CFArrayRef data = persistenceService_executeFetchRequest(request, &count, true);
        if (data) {
            for (int i = 0; i < count; ++i) {
                const id d = (const id) CFArrayGetValueAtIndex(data, i);
                int timeStrength = weekData_getWorkoutTimeForType(d, WorkoutStrength);
                time_t timestamp = weekData_getWeekStart(d);
                localtime_r(&timestamp, &localInfo);
                HistoryWeekDataModel m = {
                    .year = localInfo.tm_year % 100,
                    .month = localInfo.tm_mon,
                    .day = localInfo.tm_mday,
                    .totalWorkouts = weekData_getTotalWorkouts(d),
                    .weightArray = {
                        weekData_getLiftingLimitForType(d, LiftSquat),
                        weekData_getLiftingLimitForType(d, LiftPullup),
                        weekData_getLiftingLimitForType(d, LiftBench),
                        weekData_getLiftingLimitForType(d, LiftDeadlift)
                    },
                    .durationByType = {
                        timeStrength,
                        weekData_getWorkoutTimeForType(d, WorkoutHIC),
                        weekData_getWorkoutTimeForType(d, WorkoutSE),
                        weekData_getWorkoutTimeForType(d, WorkoutEndurance)
                    },
                    .cumulativeDuration = {[0] = timeStrength}
                };

                for (int j = 1; j < 4; ++j)
                m.cumulativeDuration[j] = m.cumulativeDuration[j - 1] + m.durationByType[j];
                array_push_back(weekData, this->data, m);
            }
        }
    }));
}

void historyViewModel_formatDataForTimeRange(HistoryViewModel *this, int index) {
    this->isSmall = true;
    this->totalWorkoutsModel.avgWorkouts = 0;
    this->totalWorkoutsModel.yMax = 0;
    this->workoutTypeModel.yMax = 0;
    this->liftModel.yMax = 0;
    memset(this->liftModel.totalByExercise, 0, 4 * sizeof(int));
    memset(this->workoutTypeModel.totalByType, 0, 4 * sizeof(int));
    array_clear(object, this->totalWorkoutsModel.entries);
    for (int i = 0; i < 4; ++i) {
        array_clear(object, this->workoutTypeModel.entries[i]);
        array_clear(object, this->liftModel.entries[i]);
    }
    array_clear(object, this->workoutTypeModel.entries[4]);

    int size = 0;
    if (!(size = (this->data->size))) return;

    int startIndex = 0;
    if (index == HistoryTimeRange6Months) {
        startIndex = size - 26;
    } else if (index == HistoryTimeRange1Year) {
        startIndex = size - 52;
    }

    if (startIndex < 0)
        startIndex = 0;
    int nEntries = size - startIndex;
    if (nEntries >= 7)
        this->isSmall = false;

    HistoryWeekDataModel *arr = this->data->arr;
    int totalWorkouts = 0, maxWorkouts = 0, maxActivityTime = 0, maxWeight = 0;

    for (int i = startIndex; i < size; ++i) {
        HistoryWeekDataModel *e = &arr[i];

        int workouts = e->totalWorkouts;
        totalWorkouts += workouts;
        if (workouts > maxWorkouts)
            maxWorkouts = workouts;
        createNewEntry(this->totalWorkoutsModel.entries, i, workouts);

        for (int j = 0; j < 4; ++j) {
            this->workoutTypeModel.totalByType[j] += e->durationByType[j];

            int weight = e->weightArray[j];
            this->liftModel.totalByExercise[j] += weight;
            if (weight > maxWeight)
                maxWeight = weight;
            createNewEntry(this->liftModel.entries[j], i, weight);
        }

        if (e->cumulativeDuration[3] > maxActivityTime)
            maxActivityTime = e->cumulativeDuration[3];
        createNewEntry(this->workoutTypeModel.entries[0], i, 0);
        for (int j = 1; j < 5; ++j)
            createNewEntry(this->workoutTypeModel.entries[j], i, e->cumulativeDuration[j - 1]);
    }

    this->totalWorkoutsModel.avgWorkouts = (float) totalWorkouts / nEntries;
    this->totalWorkoutsModel.yMax = maxWorkouts < 7 ? 7 : 1.1 * maxWorkouts;
    this->workoutTypeModel.yMax = 1.1 * maxActivityTime;
    this->liftModel.yMax = 1.1 * maxWeight;

    char buf[10];
    CFStringRef label = CFStringCreateWithFormat(NULL, NULL, this->totalWorkoutsModel.legendFormat,
                                                 this->totalWorkoutsModel.avgWorkouts);
    setLegendLabel(this->totalWorkoutsModel.legendEntries[0], label);
    CFRelease(label);

    for (int i = 0; i < 4; ++i) {
        float liftAverage = (float) this->liftModel.totalByExercise[i] / nEntries;
        int typeAverage = this->workoutTypeModel.totalByType[i] / nEntries;
        if (typeAverage > 59) {
            sprintf(buf, "%d h %d m", typeAverage / 60, typeAverage % 60);
        } else {
            sprintf(buf, "%d m", typeAverage);
        }
        label = CFStringCreateWithFormat(NULL, NULL, this->workoutTypeModel.legendFormat,
                                         this->workoutTypeModel.names[i], buf);
        setLegendLabel(this->workoutTypeModel.legendEntries[i], label);
        CFRelease(label);

        label = CFStringCreateWithFormat(NULL, NULL, this->liftModel.legendFormat,
                                         this->liftModel.names[i], liftAverage);
        setLegendLabel(this->liftModel.legendEntries[i], label);
        CFRelease(label);
    }
}

CFStringRef historyViewModel_getXAxisLabel(HistoryViewModel *this, int index) {
    CFRelease(this->formatter.currString);
    const HistoryWeekDataModel *model = &this->data->arr[index];
    this->formatter.currString = CFStringCreateWithFormat(NULL, NULL, CFSTR("%@/%d/%d"),
                                                          this->formatter.months[model->month],
                                                          model->day, model->year);
    return this->formatter.currString;
}

CFStringRef workoutTypeViewModel_getDuration(WorkoutTypeChartViewModel *this, int minutes) {
    CFRelease(this->durationStr);
    if (!minutes) {
        this->durationStr = CFStringCreateCopy(NULL, CFSTR(""));
    } else if (minutes < 60) {
        this->durationStr = CFStringCreateWithFormat(NULL, NULL, CFSTR("%dm"), minutes);
    } else {
        this->durationStr = CFStringCreateWithFormat(NULL, NULL, CFSTR("%dh %dm"),
                                                     minutes / 60, minutes % 60);
    }
    return this->durationStr;
}
